### ISSUE TITLE
Failing to restore static PIE

### ISSUE THREAD

**Issue by rafaelgieschke** [checkpoint-restore/criu#1507]
CRIU fails to restore any static PIE (ET_DYN binary without PT_INTERP segment; `gcc -static-pie`).

The attached [static-pie-test](https://github.com/checkpoint-restore/criu/files/6643746/static-pie-test.txt) script fails at https://github.com/checkpoint-restore/criu/blob/73d76d45c16f95d3f00d0331265ea04d6e54c7be/criu/pie/restorer.c#L1705 and produces [restore.log](https://github.com/checkpoint-restore/criu/files/6643752/restore.log):

```
(00.070704) pie: 1: Error (criu/pie/restorer.c:1705): prctl failed @1705 with -22
(00.070718) pie: 1: Error (criu/pie/restorer.c:1707): prctl failed @1707 with -22
(00.070723) pie: 1: Error (criu/pie/restorer.c:1708): prctl failed @1708 with -22
(00.070727) pie: 1: Error (criu/pie/restorer.c:1710): prctl failed @1710 with -22
(00.070730) pie: 1: Error (criu/pie/restorer.c:1711): prctl failed @1711 with -22
```

I think this ultimately might be considered a kernel bug as the condition in https://github.com/torvalds/linux/blob/f21b807c3cf8cd7c5ca9e406b27bf1cd2f1c1238/kernel/sys.c#L1950 does not hold (anymore), at least not after <https://github.com/torvalds/linux/commit/bbdc6076d2e5d07db44e74c11b01a3e27ab90b32>:

```
	/*
	 * @brk should be after @end_data in traditional maps.
	 */
	if (prctl_map->start_brk <= prctl_map->end_data ||
	    prctl_map->brk <= prctl_map->end_data)
		goto out;
```

The original process has /proc/[pid]/stat ([static-pie-test.log](https://github.com/checkpoint-restore/criu/files/6643749/static-pie-test.log)):

```
    45	140119223922144
    46	140119223943472
    47	93824992436224
    48	140729029304210
    49	140729029304219
    50	140729029304219
    51	140729029304303
```

with start_brk (47) < end_data (46).

If you apply [ignore.patch](https://github.com/checkpoint-restore/criu/files/6643753/ignore.patch.txt), restoring works but `syscall(SYS_brk, 0)` reports wrong values (the `sbrk(0)` glibc wrapper caches values):

```diff
--- a/criu/pie/restorer.c
+++ b/criu/pie/restorer.c
@@ -1728,6 +1728,7 @@ long __export_restore_task(struct task_restore_args *args)
 		sys_close(args->fd_exe_link);
 	}
 
+	ret = 0;
 	if (ret)
 		goto core_restore_end;
```

[ignore.log](https://github.com/checkpoint-restore/criu/files/6643754/ignore.log):

```
# before checkpoint
sbrk(0) = 0x555555b1c000
syscall(SYS_brk, 0) = 555555b1c000
# after restore
sbrk(0) = 0x555555b1c000
syscall(SYS_brk, 0) = 55b4024ce000
```

If you apply [hack.patch](https://github.com/checkpoint-restore/criu/files/6643757/hack.patch.txt), restoring and (s)brk work fine:

```diff
--- a/criu/pie/restorer.c
+++ b/criu/pie/restorer.c
@@ -1679,6 +1679,9 @@ long __export_restore_task(struct task_restore_args *args)
 	if (ret)
 		goto core_restore_end;
 
+	if (args->mm.mm_start_brk <= args->mm.mm_end_data)
+		args->mm.mm_start_data = args->mm.mm_end_data = args->mm.mm_start_brk - 1;
+
 	/*
 	 * New kernel interface with @PR_SET_MM_MAP will become
 	 * more widespread once kernel get deployed over the world.
```

[hack.log](https://github.com/checkpoint-restore/criu/files/6643758/hack.log):

```
# before checkpoint
sbrk(0) = 0x5555567ce000
syscall(SYS_brk, 0) = 5555567ce000
# after restore
sbrk(0) = 0x5555567ce000
syscall(SYS_brk, 0) = 5555567ce000
```

- [static-pie-test.txt](https://github.com/checkpoint-restore/criu/files/6643746/static-pie-test.txt)
- [static-pie-test.log](https://github.com/checkpoint-restore/criu/files/6643749/static-pie-test.log)
- [dump.log](https://github.com/checkpoint-restore/criu/files/6643751/dump.log)
- [restore.log](https://github.com/checkpoint-restore/criu/files/6643752/restore.log)
- [ignore.patch.txt](https://github.com/checkpoint-restore/criu/files/6643753/ignore.patch.txt)
- [ignore.log](https://github.com/checkpoint-restore/criu/files/6643754/ignore.log)
- [ignore-dump.log](https://github.com/checkpoint-restore/criu/files/6643755/ignore-dump.log)
- [ignore-restore.log](https://github.com/checkpoint-restore/criu/files/6643756/ignore-restore.log)
- [hack.patch.txt](https://github.com/checkpoint-restore/criu/files/6643757/hack.patch.txt)
- [hack.log](https://github.com/checkpoint-restore/criu/files/6643758/hack.log)
- [hack-dump.log](https://github.com/checkpoint-restore/criu/files/6643759/hack-dump.log)
- [hack-restore.log](https://github.com/checkpoint-restore/criu/files/6643760/hack-restore.log)

**Comment by avagin** [comment]
@0x7f454c46 could you take a look at this?

**Comment by avagin** [comment]
@rafaelgieschke I agree that this looks like a kernel bug. Would you like to send a kernel fix?

**Comment by cyrillos** [comment]
Could you please give the patch a shot?
https://lore.kernel.org/lkml/20210121221207.GB2174@grain/
